### PR TITLE
Migrate version catalog

### PR DESCRIPTION
**Pull request by AbdelrahmanEsam**
migrate to version catalog for stable dependency injection between modules

* Follow on from https://github.com/ankidroid/Anki-Android/pull/15915

### CODE REVIEW

File: AnkiDroid/build.gradle [hunk 1]
```diff
@@ -340,82 +341,82 @@ dependencies {
...
-    implementation 'com.drakeet.drawer:drawer:1.0.3'
-    implementation 'uk.co.samuelwall:material-tap-target-prompt:3.3.2'
-    implementation 'com.github.mrudultora:Colorpicker:1.2.0'
-    implementation "org.jetbrains.kotlin:kotlin-reflect:$kotlin_version"
-    implementation "org.jetbrains.kotlin:kotlin-test:$kotlin_version"
-    implementation 'com.github.ByteHamster:SearchPreference:2.5.1'
+   implementation libs.acra.limiter
+   implementation libs.acra.toast
+   implementation libs.acra.dialog
+   implementation libs.acra.http
```
Comment (david-allison) [inline AnkiDroid/build.gradle:352]: These are each misaligned with the rest of the file - add a space before each line

File: AnkiDroid/build.gradle [hunk 2]
```diff
@@ -340,82 +341,82 @@ dependencies {
...
         exclude module: "protobuf-lite"
     }
 
     androidTestImplementation project(':testlib')
 
     // May need a resolution strategy for support libs to our versions
-    androidTestImplementation "androidx.test.espresso:espresso-core:$espresso_version"
-    androidTestImplementation("androidx.test.espresso:espresso-contrib:$espresso_version") {
+    androidTestImplementation libs.espresso.core
+    androidTestImplementation(libs.espresso.contrib) {
```
Comment (david-allison) [inline AnkiDroid/build.gradle:412]: Why are these `espresso`, but the ones below `androidx.test.xxx`

File: AnkiDroid/build.gradle [hunk 3]
```diff
@@ -340,82 +341,82 @@ dependencies {
...
-    androidTestImplementation "androidx.test:core:$androidx_test_version"
-    androidTestImplementation "androidx.test.ext:junit:$androidx_test_junit_version"
-    androidTestImplementation "androidx.test:rules:$androidx_test_version"
-    androidTestImplementation "androidx.test.uiautomator:uiautomator:$uiautomator_version"
-    androidTestImplementation "org.jetbrains.kotlin:kotlin-test:$kotlin_version"
-    androidTestImplementation "org.jetbrains.kotlin:kotlin-test-junit:$kotlin_version"
-    androidTestImplementation("androidx.fragment:fragment-testing:$fragments_version")
+    androidTestImplementation libs.androidx.test.core
+    androidTestImplementation libs.androidx.junit
+    androidTestImplementation libs.test.rules
```
Comment (david-allison) [inline AnkiDroid/build.gradle:417]: why is this not `androidx`

File: AnkiDroid/build.gradle [hunk 4]
```diff
@@ -340,82 +341,82 @@ dependencies {
...
     }
-    androidTestImplementation "androidx.test:core:$androidx_test_version"
-    androidTestImplementation "androidx.test.ext:junit:$androidx_test_junit_version"
-    androidTestImplementation "androidx.test:rules:$androidx_test_version"
-    androidTestImplementation "androidx.test.uiautomator:uiautomator:$uiautomator_version"
-    androidTestImplementation "org.jetbrains.kotlin:kotlin-test:$kotlin_version"
-    androidTestImplementation "org.jetbrains.kotlin:kotlin-test-junit:$kotlin_version"
-    androidTestImplementation("androidx.fragment:fragment-testing:$fragments_version")
+    androidTestImplementation libs.androidx.test.core
+    androidTestImplementation libs.androidx.junit
```
Comment (david-allison) [inline AnkiDroid/build.gradle:416]: Why is this not `androidx.test.junit`

File: gradle/libs.versions.toml [hunk 5]
```diff
@@ -0,0 +1,178 @@
+[versions]
+acra = '5.11.3'
```
Comment (david-allison) [inline gradle/libs.versions.toml:2]: I don't feel the comments on the previous PR were addressed

The goal is to sort by use (core, test, androidTest), then alphabetically

This is just alphabetical

**Review by david-allison — CHANGES_REQUESTED** [review]

File: gradle/libs.versions.toml (hunk 5 above)
Comment (AbdelrahmanEsam) [reply gradle/libs.versions.toml:2]: sorry explain more please ?

Comment (AbdelrahmanEsam) [reply gradle/libs.versions.toml:2]: what do you mean by core ? dependencies needed in the application itself ? not test or android test ?

File: AnkiDroid/build.gradle (hunk 4 above)
Comment (AbdelrahmanEsam) [reply AnkiDroid/build.gradle:416]: it was like this before but one of the reviewers said to me that names like these is too long .... what should I do ? stick with the standards or make it as short as possible ?

File: AnkiDroid/build.gradle (hunk 3 above)
Comment (AbdelrahmanEsam) [reply AnkiDroid/build.gradle:417]: ++

File: AnkiDroid/build.gradle (hunk 2 above)
Comment (AbdelrahmanEsam) [reply AnkiDroid/build.gradle:412]: ++

File: gradle/libs.versions.toml [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+jsoup = { module = "org.jsoup:jsoup", version.ref = "jsoup" }
+kotlin-stdlib = { module = "org.jetbrains.kotlin:kotlin-stdlib", version.ref = "kotlin" }
+kotlinx-serialization-json = { module = "org.jetbrains.kotlinx:kotlinx-serialization-json", version.ref = "kotlinxSerializationJson" }
+kotlinx-coroutines-core = { module = "org.jetbrains.kotlinx:kotlinx-coroutines-core" }
+leakcanary-android = { module = "com.squareup.leakcanary:leakcanary-android", version.ref = "leakcanaryAndroid" }
+google-material = { module = "com.google.android.material:material", version.ref = "material" }
+tapTargetPrompt = { module = "uk.co.samuelwall:material-tap-target-prompt", version.ref = "materialTapTargetPrompt" }
+nanohttpd = { module = "org.nanohttpd:nanohttpd", version.ref = "nanohttpd" }
+okhttp = { module = "com.squareup.okhttp3:okhttp", version.ref = "okhttp" }
+lint-api = { module = "com.android.tools.lint:lint-api", version.ref = "lint" }
```
Comment (lukstbit) [inline gradle/libs.versions.toml:119]: Should be _android-lint-api_.

File: gradle/libs.versions.toml [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+android-lint = { module = "com.android.tools.lint:lint", version.ref = "lint" }
+android-lint-tests = { module = "com.android.tools.lint:lint-tests", version.ref = "lint" }
+hamcrest-library = { module = "org.hamcrest:hamcrest-library", version.ref = "hamcrest" }
```
Comment (lukstbit) [inline gradle/libs.versions.toml:122]: Should be in the _testing libs_, near the standard _hamcrest_ defined library.

File: gradle/libs.versions.toml [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+drawer = "1.0.3"
+dokka = "1.9.10"
+espresso = '3.5.1'
+googleAnalyticsJava7 = "2.0.13"
+hamcrest = "2.2"
+input = "3.3.0"
+javaSemver = "0.10.2"
+json = "20240205"
+jsoup = "1.17.2"
+junit = "1.1.5"
```
Comment (lukstbit) [inline gradle/libs.versions.toml:45]: This name is confusing, it's used with _androidx-test-junit_ so it should use that name _androidTestJunit_.

File: gradle/libs.versions.toml [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+androidx-sqlite-framework = { module = "androidx.sqlite:sqlite-framework", version.ref = "androidxSqliteFramework" }
+androidx-recyclerview = { module = "androidx.recyclerview:recyclerview", version.ref = "androidxRecyclerview" }
+auto-service = { module = "com.google.auto.service:auto-service", version.ref = "autoService" }
+auto-service-annotations = { module = "com.google.auto.service:auto-service-annotations", version.ref = "autoServiceAnnotations" }
+jetbrains-annotations = { module = "org.jetbrains:annotations", version.ref = "annotations" }
+colorpicker = { module = "com.github.mrudultora:Colorpicker", version.ref = "colorpicker" }
+commons-io = { module = "commons-io:commons-io", version.ref = "commonsIo" }
+commons-collections4 = { module = "org.apache.commons:commons-collections4", version.ref = "commonsCollections4" }
+commons-compress = { module = "org.apache.commons:commons-compress", version.ref = "commonsCompress" }
+core = { module = "com.afollestad.material-dialogs:core", version.ref = "androidxTextCore" }
```
Comment (lukstbit) [inline gradle/libs.versions.toml:102]: _core_ is too general, it should be _materialDialogs-core_.

File: gradle/libs.versions.toml [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+desugar-jdk-libs-nio = { module = "com.android.tools:desugar_jdk_libs_nio", version.ref = "desugar-jdk-libs-nio" }
+drakeet-drawer = { module = "com.drakeet.drawer:drawer", version.ref = "drawer" }
+google-analytics-java7 = { module = "net.mikehardy:google-analytics-java7", version.ref = "googleAnalyticsJava7" }
+input = { module = "com.afollestad.material-dialogs:input", version.ref = "input" }
```
Comment (lukstbit) [inline gradle/libs.versions.toml:106]: _input_ is too general, it should be _materialDialogs-input_(and placed near the other _materialDialogs*_ dependency).

File: gradle/libs.versions.toml [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+lint-api = { module = "com.android.tools.lint:lint-api", version.ref = "lint" }
+android-lint = { module = "com.android.tools.lint:lint", version.ref = "lint" }
+android-lint-tests = { module = "com.android.tools.lint:lint-tests", version.ref = "lint" }
+hamcrest-library = { module = "org.hamcrest:hamcrest-library", version.ref = "hamcrest" }
+protobuf-kotlin-lite = { module = "com.google.protobuf:protobuf-kotlin-lite", version.ref = "protobufKotlinLite" }
+search-preference = { module = "com.github.ByteHamster:SearchPreference", version.ref = "searchpreference" }
+seismic = { module = "com.squareup:seismic", version.ref = "seismic" }
+slf4j-timber = { module = "com.arcao:slf4j-timber", version.ref = "slf4jTimber" }
+jakewharton-timber = { module = "com.jakewharton.timber:timber", version.ref = "timber" }
+androidx-test-rules = { module = "androidx.test:rules", version.ref = "androidxTextCore" }
```
Comment (lukstbit) [inline gradle/libs.versions.toml:128]: Should be in the testing libs group of dependencies.

File: gradle/libs.versions.toml [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+androidx-work-runtime-ktx = { group = "androidx.work", name = "work-runtime-ktx", version.ref = "workRuntimeKtx" }
```
Comment (lukstbit) [inline gradle/libs.versions.toml:129]: Be consistent and use _module + version_ like all the other declared dependencies.

File: gradle/libs.versions.toml [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+
+
+#testing libs
+androidx-espresso-contrib = { module = "androidx.test.espresso:espresso-contrib", version.ref = "espresso" }
+androidx-espresso-core = { module = "androidx.test.espresso:espresso-core", version.ref = "espresso" }
+androidx-fragment-testing = { module = "androidx.fragment:fragment-testing", version.ref = "androidxFragmentKtx" }
+androidx-fragment-testing-manifest = { module = "androidx.fragment:fragment-testing-manifest", version.ref = "androidxFragmentKtx" }
+androidx-test-junit = { module = "androidx.test.ext:junit", version.ref = "junit" }
+androidx-runner = { module = "androidx.test:runner", version.ref = "androidxRunner" }
```
Comment (lukstbit) [inline gradle/libs.versions.toml:138]: androidx-test-runner

File: gradle/libs.versions.toml [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+androidx-rules = { module = "androidx.test:rules", version.ref = "androidxTextCore" }
```
Comment (lukstbit) [inline gradle/libs.versions.toml:139]: androidx-test-rules

**Review by lukstbit — CHANGES_REQUESTED** [review]

File: AnkiDroid/build.gradle [hunk 15]
```diff
@@ -340,82 +341,81 @@ dependencies {
...
+    implementation libs.acra.toast
+    implementation libs.acra.dialog
+    implementation libs.acra.http
+
+    implementation libs.materialDialogs.core
+    implementation libs.materialDialogs.input
+    implementation libs.commons.compress
+    implementation libs.commons.collections4 // SetUniqueList
+    implementation libs.commons.io // FileUtils.contentEquals
+    implementation libs.google.analytics.java7
```
Comment (david-allison) [inline AnkiDroid/build.gradle:359]: This is not a google dependency. This is a dependency for Google Analytics

File: AnkiDroid/build.gradle [hunk 16]
```diff
@@ -340,82 +341,81 @@ dependencies {
...
+    testImplementation libs.robolectric
+    testImplementation libs.androidx.test.core
+    testImplementation libs.androidx.test.junit
+    testImplementation libs.kotlin.reflect
+    testImplementation libs.kotlin.test
+    testImplementation libs.kotlin.test.junit5
+    testImplementation libs.kotlinx.coroutines.test
+    testImplementation libs.mockk
+    testImplementation libs.commons.exec // obtaining the OS
+    testImplementation(libs.androidx.fragment.testing)
```
Comment (david-allison) [inline AnkiDroid/build.gradle:396]: I don't think we need the `()` now

File: AnkiDroid/build.gradle [hunk 17]
```diff
@@ -340,82 +341,81 @@ dependencies {
...
-    testImplementation "androidx.test:runner:$androidx_test_version"
-    testImplementation "androidx.test:rules:$androidx_test_version"
-    testImplementation "androidx.test.espresso:espresso-core:$espresso_version"
-    testImplementation("androidx.test.espresso:espresso-contrib:$espresso_version") {
+    testImplementation libs.json
+    testImplementation libs.ivanshafran.shared.preferences.mock
+    testImplementation libs.androidx.test.runner
+    testImplementation libs.androidx.test.rules
+    testImplementation libs.androidx.espresso.core
+    testImplementation(libs.androidx.espresso.contrib) {
```
Comment (david-allison) [inline AnkiDroid/build.gradle:403]: is it possible to remove the `()`, or is this needed for the `{ }`

File: gradle/libs.versions.toml (hunk 5 above)
Comment (david-allison) [reply gradle/libs.versions.toml:2]: Resolving as-per https://github.com/ankidroid/Anki-Android/pull/15915#discussion_r1532644052

**Review by david-allison — CHANGES_REQUESTED** [review]
I think this is it. One name change requested

File: AnkiDroid/build.gradle (hunk 17 above)
Comment (AbdelrahmanEsam) [reply AnkiDroid/build.gradle:403]: this is needed for the { }

File: AnkiDroid/build.gradle [hunk 18]
```diff
@@ -298,39 +298,40 @@ dependencies {
...
+    implementation libs.androidx.viewpager2
+    implementation libs.androidx.constraintlayout
+    implementation libs.androidx.webkit
+    implementation libs.google.material
+    implementation libs.android.image.cropper
+    implementation libs.nanohttpd
+    implementation libs.kotlinx.serialization.json
+    implementation libs.seismic
+
+    debugImplementation(libs.androidx.fragment.testing.manifest)
```
Comment (david-allison) [inline AnkiDroid/build.gradle:330]: `()` probably isn't needed

File: AnkiDroid/build.gradle (hunk 16 above)
Comment (david-allison) [reply AnkiDroid/build.gradle:396]: Why was this resolved?

File: settings.gradle.kts [hunk 19]
```diff
@@ -26,4 +26,4 @@ dependencyResolutionManagement {
     }
 }
 
-include(":lint-rules", ":api", ":AnkiDroid", ":testlib")
+include(":lint-rules", ":api", ":AnkiDroid", ":testlib")
```
Comment (david-allison) [inline settings.gradle.kts:29]: could you revert this file

**Review by david-allison — APPROVED** [review]
Only nitpicks, cheers!

**Review by lukstbit — APPROVED** [review]
LGTM, thank you for contributing!

**Pull request merged**